[JIT] Add default arguments for class types

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45098 [JIT] Add default arguments for class types**

**Summary**
This commit adds support for default arguments in methods of class
types. Similar to how default arguments are supported for regular
script functions and methods on scripted modules, default values are
retrieved from the definition of a TorchScript class in Python as Python
objects, converted to IValues, and then attached to the schemas of
already compiled class methods.

**Test Plan**
This commit adds a set of new tests to TestClassType to test default
arguments.

**Fixes**
This commit fixes #42562.

Differential Revision: [D23844769](https://our.internmc.facebook.com/intern/diff/D23844769)

Commit:
[JIT] Add default arguments for class types

**Summary**
This commit adds support for default arguments in methods of class
types. Similar to how default arguments are supported for regular
script functions and methods on scripted modules, default values are
retrieved from the definition of a TorchScript class in Python as Python
objects, converted to IValues, and then attached to the schemas of
already compiled class methods.

**Test Plan**
This commit adds a set of new tests to TestClassType to test default
arguments.

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

SplitInfinity commented:
This is a different approach to #42562 than #42988 that requires less changes but uses the same approach that we use for methods and functions (take Python objects representing the default values, convert them to IValues, and attach them to the schemas after compilation).

eellison commented on torch/jit/frontend.py:
```diff
@@ -305,6 +305,32 @@ def get_default_args(fn):
+
+    # Get method defaults. Property defaults do not need to be considered
+    # because setters cannot be invoked without a value.
+    defaults = {method[0]: get_default_args(method[1]) for method in methods}
```
nit maybe: `defaults = {method_name: get_default_args(method_impl)) for method_name, method_impl in methods}` to be a little more readable

SplitInfinity replied:
Woah, I didn't know you unpack tuples like that inside a comprehension but it makes sense. Cool!

eellison reviewed (approved):
🚢

Commit:
Update on "[JIT] Add default arguments for class types"


**Summary**
This commit adds support for default arguments in methods of class
types. Similar to how default arguments are supported for regular
script functions and methods on scripted modules, default values are
retrieved from the definition of a TorchScript class in Python as Python
objects, converted to IValues, and then attached to the schemas of
already compiled class methods.

**Test Plan**
This commit adds a set of new tests to TestClassType to test default
arguments.

**Fixes**
This commit fixes #42562.

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1339,6 +1341,29 @@ void initJitScriptBindings(PyObject* module) {
+          auto def_name = (*defs_it).name().name();
+          // If the method is not in the defaults map, assume there are
+          // no default arguments for it.
+          if (defaults.find(def_name) == defaults.end()) {
```
nit:

save results of `defaults.find(def_name)` and reuse on line 1364 to look up only once

SplitInfinity replied:
Nice catch.

gmagogsfm commented on torch/jit/frontend.py:
```diff
@@ -305,6 +305,32 @@ def get_default_args(fn):
+        A Dict[str, Dict[str, Any]] which maps each method name to a Dict[str, Any]
+        that maps each argument name to its default value.
+    """
+    # Get methods (except static methods because those are compiled separately as
```
OOC, are default arg values for static methods already supported?

SplitInfinity replied:
Yeah, static methods of classes are actually compiled as if they were standalone methods (since there is no `self` argument or associated instance).

gmagogsfm commented on torch/jit/frontend.py:
```diff
@@ -305,6 +305,32 @@ def get_default_args(fn):
+        cls,
+        predicate=lambda m: (inspect.ismethod(m) or inspect.isfunction(m))
+        and not is_static_fn(cls, m.__name__)
+        and m.__name__ in cls.__dict__
```
Could you shed some light on why needing to check `m.__name__ in cls.__dict__`? in case user doesn't want to expose this method?

SplitInfinity replied:
To be honest I copied this from `get_jit_class_def`.

Commit:
Update on "[JIT] Add default arguments for class types"


**Summary**
This commit adds support for default arguments in methods of class
types. Similar to how default arguments are supported for regular
script functions and methods on scripted modules, default values are
retrieved from the definition of a TorchScript class in Python as Python
objects, converted to IValues, and then attached to the schemas of
already compiled class methods.

**Test Plan**
This commit adds a set of new tests to TestClassType to test default
arguments.

**Fixes**
This commit fixes #42562.

Differential Revision: [D23844769](https://our.internmc.facebook.com/intern/diff/D23844769)

[ghstack-poisoned]
